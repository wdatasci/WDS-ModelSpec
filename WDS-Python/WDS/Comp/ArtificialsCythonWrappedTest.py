#!/usr/bin/env python3


import ArtificialsCythonWrapped as art_c

import numpy as np

np.set_printoptions(linewidth=np.inf)

import polars as pl
pl.cfg.Config.set_tbl_cols(20)
pl.cfg.Config.set_tbl_rows(100)


if __name__=="__main__":
    #print(dir(art_c))
    #print(art_c.__name__)
    s="Buckets"
    print("s=",s," clean treatment=",art_c._eTreatmentClean(s))
    X=np.ndarray((41,1),dtype=np.double)
    for i in range(0,41):
        X[i,0]=(i-20.0)/2.0;
    X[2,0]=np.nan
    X[10,0]=np.nan

    Y=['']*41 #np.ndarray((41,1),dtype=str)
    for i in range(0,41):
        Y[i]=chr(65+i)
    Y[2]=''
    Y[10]=''
    Y=pl.DataFrame({'y':Y})

    for iT,T in enumerate(["Hats","Levels","DiscreteRC","DiscreteLC","iHats","BZ2","BZ3"]):

        CriticalValues=np.array([[ -6, -1, 3, 6, 8, 9 ]],np.double)
        CleanLimits=np.array([[ -8, 10.5 ]],np.double)
        nArtificials=art_c._nArtificialCount(CriticalValues.shape[1],T)
        print("X=")
        print(X)
        print("T=",T)
        print("CriticalValues=",CriticalValues)
        print("CleanLimits=",CleanLimits)
        print("nArtificials=",nArtificials)
        XA=art_c.fArtificials(X,T,CriticalValues,CleanLimits)
        print("Labels=",art_c.fArtificialLabels(CriticalValues.shape[1],T))
        print("XA=",XA)
        Coef=2.0-4.0*np.random.rand(3,nArtificials)
        print("Coef=",Coef,type(Coef))
        XAS=art_c.fArtificialsScored(X,T,CriticalValues,CleanLimits=CleanLimits,CoefficientsSet=Coef)
        print("Labels=",art_c.fArtificialsScoredLabels(Coef.shape[0]))
        print("XAS=",XAS)
        print(type(XAS))

    print("Y=")
    print(Y)
    print("Treatment='Categorical' with CriticalValues="+str([['A','D'],['e','d'],['X','Y','Z','G','x','y']]))
    print("Treatment='Categorical' with CriticalValues="+str([['A','D'],['e','d'],['X','Y','Z','G','x','y']]))
    YA=art_c.fArtificials(Y.to_numpy(),'Categorical',[['A','D'],['e','d'],['X','Y','Z','G','x','y']])
    print("YA=",YA)
    print(type(YA))

    Y=Y.hstack(pl.DataFrame(YA))
    print(Y)




