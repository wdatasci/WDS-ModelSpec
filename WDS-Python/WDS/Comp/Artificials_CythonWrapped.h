/* Generated by Cython 0.29.26 */

#ifndef __PYX_HAVE__Artificials_CythonWrapped
#define __PYX_HAVE__Artificials_CythonWrapped

#include "Python.h"

#ifndef __PYX_HAVE_API__Artificials_CythonWrapped

#ifndef __PYX_EXTERN_C
  #ifdef __cplusplus
    #define __PYX_EXTERN_C extern "C"
  #else
    #define __PYX_EXTERN_C extern
  #endif
#endif

#ifndef DL_IMPORT
  #define DL_IMPORT(_T) _T
#endif

__PYX_EXTERN_C PyObject *_eTreatmentClean(PyObject *);

#endif /* !__PYX_HAVE_API__Artificials_CythonWrapped */

/* WARNING: the interface of the module init function changed in CPython 3.5. */
/* It now returns a PyModuleDef instance instead of a PyModule instance. */

#if PY_MAJOR_VERSION < 3
PyMODINIT_FUNC initArtificials_CythonWrapped(void);
#else
PyMODINIT_FUNC PyInit_Artificials_CythonWrapped(void);
#endif

#endif /* !__PYX_HAVE__Artificials_CythonWrapped */
