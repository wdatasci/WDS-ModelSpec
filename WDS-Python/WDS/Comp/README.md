# Artificials

To compare both Cython, raw, and JIT versions of the Artificials.h C code

* Build
    This assumes that cython and python3 distributions and dependencies are available.

    Build with:  
        (rm __init__.py if it tries to install in a WDS subdirectory)
        export CPPFLAGS=-I/usr/local/lib/python3.9/dist-packages/numpy/core/include
        python3 setup.py build_ext --inplace; python3 setup.py clean --all
        (echo "pass" > __init__.py if necessary)

