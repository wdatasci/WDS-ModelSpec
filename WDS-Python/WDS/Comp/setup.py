from setuptools import setup

from Cython.Build import cythonize

setup(ext_modules=cythonize("ArtificialsCythonWrapped.pyx",language_level="3"))

