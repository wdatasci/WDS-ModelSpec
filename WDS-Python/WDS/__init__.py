'''A general parent module for Wypasek Data Science, Inc. modules
'''


