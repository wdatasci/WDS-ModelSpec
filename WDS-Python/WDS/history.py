'''
Unix-like history and ! during a python session.

Usage:
    from WDS.history import history_init

    #instantiate link to global namespace with

    history_init(globals())

    history() # to list history
    history(last=10) # to list the last 10 lines of history

    hbang(i) # as in shell, !### will run history line ###
             #  hbang(i) will evaluate history line i in the global namespace

'''
#import sys
from WDS.namespaceop import namespaceop
#_using__dict__

#if hasattr(sys,'ps1'):
import readline
#_using__dict__
@namespaceop
def history_init(arg):
    __history_namespace = locals()
    def history(last=None):
        n=readline.get_current_history_length()
        start = n-last if (last is not None) else 0
        for i in range(start,n):
            print( i, readline.get_history_item(i) )
    def hbang(i=None):
        global __history_namespace
        readline.add_history(readline.get_history_item(i))
        eval(compile(readline.get_history_item(i),"history line "+str(i),"exec"),globals(),__history_namespace)
#else:
#    @namespaceop_using__dict__
#    def history_init(arg):
#        def history(arg): pass
#        def hbang(arg): pass
#
