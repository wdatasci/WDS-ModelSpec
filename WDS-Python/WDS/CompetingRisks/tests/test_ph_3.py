
import numpy as np
import numpy.random as rand
import polars as pl
import math
from typing import *

import WDS.Comp.ArtificialsCythonWrapped as art_c

from WDS.Comp.WDSModelPrep import *

from WDS.history import *
history_init(globals())

#build a model spec from scratch

def Bones_explode(df : pl.DataFrame,
        /,
        RowCountName : str = "N",
        RowIndexName : str = 'RowIndex',
        OffsetName : str = 'RowIndex_First',
        ) -> pl.DataFrame:
    rv = df.select([pl.col('*'),pl.concat_list([pl.col(OffsetName),pl.col(RowCountName)]).alias(RowIndexName)])
    rv = rv.with_column(pl.col(RowIndexName).apply(lambda x:[i+x[0] for i in range(0,x[1]+1)])).explode(RowIndexName)
    return rv

def Bones(ID_List : List[int],
        N_List   : List[int],
        Offset_List  :   List[int],
        /,
        IDName : str = 'ID',
        RowIndexName : str = 'RowIndex',
        OffsetName : str = 'RowIndex_First',
        ) -> pl.DataFrame:
    rv = pl.DataFrame({IDName:ID_List, 'N':N_List, OffsetName:Offset_List})
    return Bones_explode(rv, RowIndexName=RowIndexName, OffsetName=OffsetName)

MonthID_Start = (2020-2000)*12+1
MonthID_Stop = (2022-2000)*12+5

M = 1000

T = 100

IDs=[x for x in range(1, M+1)]

data = pl.DataFrame({'ID':IDs,
            'N':[T for x in IDs],
            'VintageMonthID':[MonthID_Start+rand.randint(MonthID_Start,MonthID_Stop) for x in IDs],
            'Strata':rand.randint(0,2,M)+1,
            'Static_A':rand.rand(M),
            'Static_B':rand.rand(M)*100-30,
            })


data2 = Bones_explode(data, 
            RowCountName='N',
            RowIndexName='MonthID', 
            OffsetName="VintageMonthID",
            )

data2['TV_A'] = rand.rand(data2.shape[0])
data2['TV_B'] = 3*rand.rand(data2.shape[0])-2

data2['RowIndex'] = data2.MonthID-data2.VintageMonthID
data2['Age'] = data2.MonthID-data2.VintageMonthID

data2['TestBaseline_1'] = data2.Age.apply(lambda x: np.exp(-np.power(x-10.0,2.0)/20.0)/100.0+x/1000.0+.002)
data2['TestBaseline_2'] = data2.Age.apply(lambda x: np.exp(-np.power(x-20.0,2.0)/40.0)/100.0+x/500.0+.001)



mdl = WDSModel("test_py_3", Responses=['Resp1', 'Resp2'])

vStatic_A = add_Variable(Model=mdl, Name='Static_A', CriticalValues=[.2, .6], CleanLimits=[.05, .95], Treatment='Hats', Coefficients=[[.3, 0, .8],[0.0, 0, 0.0]])
vStatic_A.add_DropIndex(1)
vStatic_A.VariableModelDirectives.Static = "Yes"
vStatic_A.VariableModelDirectives.ResponseUse = "Resp1"

vStatic_B= add_Variable(Model=mdl, Name='Static_B', CriticalValues=[-2, 20], CleanLimits=[-10, 40], Treatment='Hats', Coefficients=[[.3, .1, 0.0],[-.3, -.2, 0.0]])
vStatic_B.add_DropIndex(2)
vStatic_B.VariableModelDirectives.Static = "Yes"
vStatic_B.VariableModelDirectives.ResponseUse = "All"


vTV_A = add_Variable(Model=mdl, Name='TV_A', CriticalValues=[.2, .6], CleanLimits=[.05, .95], Treatment='Hats', Coefficients=[[.3, 0, .8],[-.3, 0, .17]])
vTV_A.add_DropIndex(1)
vTV_A.VariableModelDirectives.ResponseUse = "All"

vTV_B= add_Variable(Model=mdl, Name='TV_B', CriticalValues=[-2, 20], CleanLimits=[-10, 40], Treatment='Hats', Coefficients=[[0.0, 0.0, 0.0],[-.3, -.2, 0]])
vTV_B.add_DropIndex(2)
vTV_B.VariableModelDirectives.ResponseUse = "Resp2"


mdl_baseline = WDSModel("test_py_3_baseline", Responses=['All',])
vAge = add_Variable(Model=mdl_baseline, Name='Age', CriticalValues=[0.0, 10.0, 20.0, 80.0], CleanLimits=[-1.0, 140.0 ], Treatment='Hats', Coefficients=[[0.0, 0.0, 0.1, 0.08, 0.2]])
vAge.add_DropIndex(0)
vAge.add_DropIndex(1)




ModelBuildingMD = ModelBuildingMD_Constructor()(mdl)
arts_names = []
arts_droplist = []

baseline_arts = plDF(art_c.fArtificials(data2.Age+0.0,vAge.Treatment,vAge.CriticalValues.as_numpy(),vAge.CleanLimits.as_numpy(),LabelBase=vAge.Name))
baseline = plDF(art_c.fArtificialsScored(data2.Age+0.0,vAge.Treatment,vAge.CriticalValues.as_numpy(),vAge.CleanLimits.as_numpy(),vAge.CoefficientsSet.as_numpy(), LabelBase = vAge.Name + '_Marg'))

arts_names.extend( baseline_arts.columns )
arts_droplist.extend( [baseline_arts.columns[i] for i in vAge.DropIndexs.as_list()] )


Static_A_arts = fArtificials(data2.Static_A,vStatic_A)
Static_A_marg = fArtificialsScored(data2.Static_A,vStatic_A)

ModelBuildingMD.add_arts(vStatic_A, Static_A_arts.columns)
system_matrix = copy.copy(Static_A_arts)


Static_B_arts = fArtificials(data2.Static_B,vStatic_B)
Static_B_marg = fArtificialsScored(data2.Static_B,vStatic_B)
    
ModelBuildingMD.add_arts(vStatic_B, Static_B_arts.columns)
system_matrix = system_matrix.hstack(Static_B_arts)

TV_A_arts = fArtificials(data2.TV_A,vTV_A)
TV_A_marg = fArtificialsScored(data2.TV_A,vTV_A)

ModelBuildingMD.add_arts(vTV_A, TV_A_arts.columns)
system_matrix = system_matrix.hstack(TV_A_arts)

TV_B_arts = fArtificials(data2.TV_B,vTV_B)
TV_B_marg = fArtificialsScored(data2.TV_B,vTV_B)

ModelBuildingMD.add_arts(vTV_B, TV_B_arts.columns)
system_matrix = system_matrix.hstack(TV_B_arts)

ebz = ((Static_A_marg + Static_B_marg + TV_A_marg + TV_B_marg)).with_columns(pl.col('*').apply(np.exp))  #*baseline[:,0]

data2['Haz_A'] = ebz[:,0]*baseline[:,0]
data2['Haz_B'] = ebz[:,1]*baseline[:,0]

data2['eps_A'] = rand.rand(data2.shape[0])
data2['eps_B'] = rand.rand(data2.shape[0])

data2['Signal_A']=data2.eps_A<data2.Haz_A
data2['Signal_B']=data2.eps_B<data2.Haz_B

data2=data2.with_column(pl.concat_list(['Signal_A','Signal_B']).apply(lambda x: 1 if x[0] else 2 if x[1] else 0).alias('Signal'))

def f(arg):
    found=False
    for i,v in enumerate(arg):
        if v!=0:
            found=True
            break
    return i

x=data2.select(['ID','Signal']).groupby('ID').agg_list().select(['ID','Signal',pl.col('Signal').apply(f).alias('EventIndex')])
x=x.select(['ID','EventIndex',pl.concat_list(['Signal','EventIndex']).apply(lambda x:x[x[-1]]).alias('EventClass')])

data2=data2.join(x,on=['ID']).filter(pl.col('RowIndex') <= pl.col('EventIndex'))

data2_Resp1 = system_matrix.select(ModelBuildingMD.effective_names('Resp1'))
data2_Resp2 = system_matrix.select(ModelBuildingMD.effective_names('Resp2'))


fid = open(__file__+'.out','w')

mdl.export(fid, level=0)

fid.close()

vrbl = mdl.get_Variable('Static_A')

x1 = art_c.fArtificials(data2.Static_A,'Hats',vrbl.CriticalValues.as_numpy(),vrbl.CleanLimits.as_numpy())





