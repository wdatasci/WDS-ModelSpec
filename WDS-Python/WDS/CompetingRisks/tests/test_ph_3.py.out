<Model xmlns:wds="https://github.com/wdatasci/WDS-ModelSpec" Name="test_py_3" Handle="test_py_3">
    <ModelDirectives>
        <Responses>
            <Response Position="0">Resp1</Response>
            <Response Position="1">Resp2</Response>
        </Responses>
    </ModelDirectives>
    <Variable Name="Static_A" Handle="Static_A">
        <VariableModelDirectives ResponseUse="Resp1" Static="Yes"/>
        <Treatment>Hats</Treatment>
        <Source>Static_A</Source>
        <CleanLimits>
            <CleanLimit Position="0">0.05</CleanLimit>
            <CleanLimit Position="1">0.95</CleanLimit>
        </CleanLimits>
        <CriticalValues>
            <CriticalValue Position="0">0.2</CriticalValue>
            <CriticalValue Position="1">0.6</CriticalValue>
        </CriticalValues>
        <CoefficientsSet>
            <Coefficients Response="Resp1">
                <Coefficient Position="0">0.3</Coefficient>
                <Coefficient Position="1">0.1</Coefficient>
                <Coefficient Position="2">0.8</Coefficient>
            </Coefficients>
            <Coefficients Response="Resp2">
                <Coefficient Position="0">-0.3</Coefficient>
                <Coefficient Position="1">-0.2</Coefficient>
                <Coefficient Position="2">0.17</Coefficient>
            </Coefficients>
        </CoefficientsSet>
    </Variable>
    <Variable Name="Static_B" Handle="Static_B">
        <VariableModelDirectives ResponseUse="Resp2" Static="Yes"/>
        <Treatment>Hats</Treatment>
        <Source>Static_B</Source>
        <CleanLimits>
            <CleanLimit Position="0">-10</CleanLimit>
            <CleanLimit Position="1">40</CleanLimit>
        </CleanLimits>
        <CriticalValues>
            <CriticalValue Position="0">-2</CriticalValue>
            <CriticalValue Position="1">20</CriticalValue>
        </CriticalValues>
        <CoefficientsSet>
            <Coefficients Response="Resp1">
                <Coefficient Position="0">0.3</Coefficient>
                <Coefficient Position="1">0.1</Coefficient>
                <Coefficient Position="2">0.8</Coefficient>
            </Coefficients>
            <Coefficients Response="Resp2">
                <Coefficient Position="0">-0.3</Coefficient>
                <Coefficient Position="1">-0.2</Coefficient>
                <Coefficient Position="2">0.17</Coefficient>
            </Coefficients>
        </CoefficientsSet>
    </Variable>
    <Variable Name="TV_A" Handle="TV_A">
        <VariableModelDirectives ResponseUse="All"/>
        <Treatment>Hats</Treatment>
        <Source>TV_A</Source>
        <CleanLimits>
            <CleanLimit Position="0">0.05</CleanLimit>
            <CleanLimit Position="1">0.95</CleanLimit>
        </CleanLimits>
        <CriticalValues>
            <CriticalValue Position="0">0.2</CriticalValue>
            <CriticalValue Position="1">0.6</CriticalValue>
        </CriticalValues>
        <CoefficientsSet>
            <Coefficients Response="Resp1">
                <Coefficient Position="0">0.3</Coefficient>
                <Coefficient Position="1">0.1</Coefficient>
                <Coefficient Position="2">0.8</Coefficient>
            </Coefficients>
            <Coefficients Response="Resp2">
                <Coefficient Position="0">-0.3</Coefficient>
                <Coefficient Position="1">-0.2</Coefficient>
                <Coefficient Position="2">0.17</Coefficient>
            </Coefficients>
        </CoefficientsSet>
    </Variable>
    <Variable Name="TV_B" Handle="TV_B">
        <VariableModelDirectives ResponseUse="Resp2"/>
        <Treatment>Hats</Treatment>
        <Source>TV_B</Source>
        <CleanLimits>
            <CleanLimit Position="0">-10</CleanLimit>
            <CleanLimit Position="1">40</CleanLimit>
        </CleanLimits>
        <CriticalValues>
            <CriticalValue Position="0">-2</CriticalValue>
            <CriticalValue Position="1">20</CriticalValue>
        </CriticalValues>
        <CoefficientsSet>
            <Coefficients Response="Resp1">
                <Coefficient Position="0">0.3</Coefficient>
                <Coefficient Position="1">0.1</Coefficient>
                <Coefficient Position="2">0.8</Coefficient>
            </Coefficients>
            <Coefficients Response="Resp2">
                <Coefficient Position="0">-0.3</Coefficient>
                <Coefficient Position="1">-0.2</Coefficient>
                <Coefficient Position="2">0.17</Coefficient>
            </Coefficients>
        </CoefficientsSet>
    </Variable>
</Model>
